### ISSUE TITLE
Add configurable commit mode?

### ISSUE THREAD

**Issue by zejji** [dbeaver/dbeaver#1562]
Currently DBeaver supports auto-commit and manual commit modes. Would it be possible to add a configurable commit mode which only starts a transaction when certain types of queries are executed (in particular I am thinking of data-modifying queries such as INSERT, REPLACE, UPDATE and DELETE, as well as schema changes). The user could have the option of specifying which commands should automatically trigger the start of a transaction.

I think this would also solve the oft-raised usability issue of SELECT statements needing to be committed when in manual commit mode. It's a bit unintuitive (and generally not wanted) to start a transaction immediately on connecting to a database. (I certainly found this very confusing at first when browsing a production database's data and wondering why no updates were coming in!)

Obviously, any SELECT statements executed after a transaction had been triggered would still need to be committed, but perhaps there could be a clearer visual indicator that a transaction was in progress (e.g. a popup when the transaction starts, with a "don't show again option" for experienced users)?

Btw - I also wanted to say what a great program DBeaver is. It's by far and away the best database client I've used. Thanks for all your hard work!

**Comment by serge-rider** [comment]
I think this could be a handy improvement. 
Having open transaction for select queries may be annoying.

**Comment by MatheusNaldi** [comment]
I need this :+1:

**Comment by MatheusNaldi** [comment]
+1

**Comment by kartoffelsup** [comment]
The other way would be less error prone imho. Configuring statements that don't need a transaction. Otherwise strong : thumbsup:

**Comment by Mudshovel** [comment]
This is something that I've been waiting for a while.

+1

**Comment by titopepper** [comment]
Agree this will be helpful.

**Comment by NYP4N** [comment]
+1

**Comment by jmaynier** [comment]
+1 
The lack of this feature make Dbeaver less useful than other SQL IDE like Datagrip. Lots of non dev sql users don't understand why they have to rollback when they execute a select with a type or logical error

**Comment by serge-rider** [comment]
@jmaynier I think you mean PostgreSQL-specific behavior when any SQL error blocks any further SQL queries until you rollback (or commit) current transaction.

**Comment by stantheman77** [comment]
+1!!!

**Comment by wkzhu** [comment]
+1
Even double clicking a table just to view it requires commitment

**Comment by WicloW** [comment]
when this will be done?

**Comment by kseniiaguzeeva** [comment]
We are going to do this improvement in 6.3 version. Thank you for remembering this.

**Comment by serge-rider** [comment]
Problem with PostgreSQL "aborted transaction" was resolved in 6.2.something.
Problem with SELECT queries initiating transaction  monitor was solved too.
Moving the rest to 6.3

**Comment by Bschitter** [comment]
+1 love to see this feature

**Comment by serge-rider** [comment]
Generally it is working:
![image](https://user-images.githubusercontent.com/6398845/71542137-c3bad280-2973-11ea-8cec-d0af058f48a1.png)

Configuring keywords which trigger transaction mode will be implemented as a part of #6528

**Comment by p91paul** [comment]
which keywords are currently hardcoded? (the answer to this question should better go in a wiki page actually)

**Comment by serge-rider** [comment]
INSERT,DELETE,UPDATE,MERGE,UPSERT,TRUNCATE in the first alpha version.

Although TRUNCATE is often non-transactional.

I'll create wiki as soon as this feature will be ready. Perhaps some keywords customization will be added in 6.3.2.

**Comment by serge-rider** [comment]
Also CREATE,DROP,ALTER (for databases which support transactions for DDL).

**Comment by p91paul** [comment]
what about call/execute?

**Comment by p91paul** [comment]
also begin/end blocks might call stored procedures

**Comment by serge-rider** [comment]
Done. We'll check all DML statements (INSERT, UPDATE, etc), all execute statements and all block statements.
Statements customization will be implemented in #6528. Till then we'll use db-dialect-specific keywords.

**Comment by kseniiaguzeeva** [comment]
Verified

**Comment by Bschitter** [comment]
Thanks for implementing this feature.

However it seems not to work properly with a PostgreSQL database.
If I have selected "Manual Commit" and "Smart auto-commit" and execute a simple select statement, it will still result in an "idle in transaction" seen in the pg_stat_activity view. As SQL statement it shows "SHOW search_path" and as application_name "DBeaver 6.3.2 - SQLEditor <Script.sql>". Seems to be something DBeaver does for PostgreSQL databases.

While testing this I have noticed that when a PostgreSQL connection is marked as production and is therefore set to "Manual Commit" when opening a connection, there will also be an "idle in transaction" with application_name "DBeaver 6.3.2 - Main" and query:
```
FROM pg_catalog.pg_type t
LEFT OUTER JOIN pg_class c ON c.oid=t.typrelid
WHERE typnamespace=$1
ORDER by t.oid
```

PostgreSQL version: 11.5
DBeaver version: 6.3.2 (Windows 64-Bit)

**Comment by serge-rider** [comment]
@Bschitter I think it is the same as #7402 
It is not related to smart commit mode. It is how PG JDBC driver behaves.
Let's track this issue in #7402

**Comment by serge-rider** [comment]
@Bschitter after some testing I can no longer reproduce this.
Whenever you are in manual commit mode Smart auto-commit doesn't make sense (as you already in transaction).
Smart auto-commit works only if you are in auto-commit mode. It switches to Manual commit before you execute any data modifying query.
I think this is how it was supposed to work.
WDYT?

**Comment by Bschitter** [comment]
@serge-rider sorry, then I missunderstood the feature slightly. Because of your posted screenshot I thought it would be the other way around, that you are in manual commit mode and then have smart auto-commit selected and it always auto commits except if a defined modifying statements is present.

However if I am in auto commit mode with smart auto-commit selected my modifying statement (e.g `update TABLE set COLUMN = 'New Name' where id = 1;`) is instantly committed.

**Comment by serge-rider** [comment]
@Bschitter From my perspective smart auto commit shouldn't perform commit implicitly.

In your case whenever you execute a select query (being in manual txn mode) it must switch to auto-commit. But this means it must commit already opened transaction. This doesn't seem right.

> However if I am in auto commit mode with smart auto-commit selected my modifying statement (e.g update TABLE set COLUMN = 'New Name' where id = 1;) is instantly committed.

This is a bug then. Although I can't reproduce it. Does it switch to manual mode at all? If yes - does it switch after query execute?
What are your settings (is "separate connection for each editor" on?)

**Comment by serge-rider** [comment]
@Bschitter Well, I found the bug. It checks query text in case-sensitive mode.
Will be fixed in the next version, thanks for the bug report.

**Comment by Bschitter** [comment]
@serge-rider indeed it seems to be case sensitive. Now it works, thanks.
And thank you for the support and the quick fix!

**Comment by haugs** [comment]
Thanks for implementing this great feature! We recently had a lot of trouble with "idle in transaction" sessions in our productive database that we traced back to the support group using DBeaver in manual transaction mode.

The new feature looks very promising and I did some tests. I'd like to suggest two bugfixes/improvements:
1. It works fine for sessions associated with a SQL Editor. But when I double-click a table in the database navigator, the new tab does not have the "Smart auto-commit" option in its GUI at all. The feature should be available for the "Main" session too.
My people use the visual navigation and data queriyng quite a lot and would need the smart auto-commit in that environment too.

2. When you use the "Auto-Commit plus Smart auto-commit" setting it switches into manual commit mode once you use an UPDATE statement. When you then hit ROLLBACK (or COMMIT) it would be great if it would switch back to Auto-Commit as now there are no more executed modifying statements.
